boilerplate/app/components/Screen.tsx

```diff
@@ -0,0 +1,243 @@
+}
+
+interface FixedScreenProps extends BaseScreenProps {
+  preset?: "fixed"
+}
+interface ScrollScreenProps extends BaseScreenProps {
+  preset?: "scroll"
+  /**
+   * Should keyboard persist on screen tap. Defaults to handled.
+   * Only applies to scroll preset.
+   */
+  keyboardShouldPersistTaps?: "handled" | "always" | "never"
+  /**
+   * Pass any additional props directly to the ScrollView component.
+   */
+  ScrollViewProps?: ScrollViewProps
+}
+
+interface AutoScreenProps extends Omit<ScrollScreenProps, "preset"> {
+  preset?: "auto"
+  /**
+   * Threshold to trigger the automatic disabling/enabling of scroll ability.
+   * Defaults to `{ percent: 0.92 }`.
+   */
+  scrollEnabledToggleThreshold?: { percent?: number; point?: number }
+}
+
+export type ScreenProps = ScrollScreenProps | FixedScreenProps | AutoScreenProps
+
+const isIos = Platform.OS === "ios"
+
+function isNonScrolling(preset?: ScreenProps["preset"]) {
+  return !preset || preset === "fixed"
+}
+
+function useAutoPreset(props: AutoScreenProps) {
+  const { preset, scrollEnabledToggleThreshold } = props
+  const { percent = 0.92, point = 0 } = scrollEnabledToggleThreshold || {}
+
+  const scrollViewHeight = useRef(null)
+  const scrollViewContentHeight = useRef(null)
+  const [scrollEnabled, setScrollEnabled] = useState(true)
+
+  function updateScrollState() {
+    if (scrollViewHeight.current === null || scrollViewContentHeight.current === null) return
+
+    // check whether content fits the screen then toggle scroll state according to it
+    const contentFitsScreen = (function () {
+      if (!!point) {
+        return scrollViewContentHeight.current < scrollViewHeight.current - point
+      } else {
+        return scrollViewContentHeight.current < scrollViewHeight.current * percent
+      }
+    })()
+
+    // content is less than the size of the screen, so we can disable scrolling
+    if (scrollEnabled && contentFitsScreen) setScrollEnabled(false)
+
+    // content is greater than the size of the screen, so let's enable scrolling
+    if (!scrollEnabled && !contentFitsScreen) setScrollEnabled(true)
+  }
+
+  function onContentSizeChange(w: number, h: number) {
+    // update scroll-view content height
+    scrollViewContentHeight.current = h
+    updateScrollState()
+  }
+
+  function onLayout(e: LayoutChangeEvent) {
+    const { height } = e.nativeEvent.layout
+    // update scroll-view  height
+    scrollViewHeight.current = height
+    updateScrollState()
+  }
+
+  // update scroll state on every render
+  if (preset === "auto") updateScrollState()
+
+  return {
+    scrollEnabled: preset === "auto" ? scrollEnabled : true,
+    onContentSizeChange,
+    onLayout,
+  }
+}
+
+function ScreenWithoutScrolling(props: ScreenProps) {
+  const { style, contentContainerStyle, children } = props
+  return (
+    <View style={[$outerStyle, style]}>
+      <View style={[$innerStyle, contentContainerStyle]}>{children}</View>
+    </View>
+  )
+}
+
```

Comment:
Cleans up both `ScreenWithoutScrolling` and `ScreenWithScrolling` components. They used to have a lot of duplicated code.